nce/NCE-Pairwise-SM/train.py

```diff
@@ -0,0 +1,420 @@
+        model = torch.load(args.resume_snapshot, map_location=lambda storage, location: storage)
+else:
+    model = SmPlusPlus(config)
+    model.static_question_embed.weight.data.copy_(QUESTION.vocab.vectors)
+    model.nonstatic_question_embed.weight.data.copy_(QUESTION.vocab.vectors)
+    model.static_answer_embed.weight.data.copy_(ANSWER.vocab.vectors)
+    model.nonstatic_answer_embed.weight.data.copy_(ANSWER.vocab.vectors)
+
+    if args.cuda:
+        model.cuda()
+        print("Shift model to GPU")
+
+
+pw_model = PairwiseConv(model)
+
+parameter = filter(lambda p: p.requires_grad, pw_model.parameters())
+
+# the SM model originally follows SGD but Adadelta is used here
+optimizer = torch.optim.Adadelta(parameter, lr=args.lr, weight_decay=args.weight_decay, eps=1e-6)
+# A good lr is required to use Adam
+# optimizer = torch.optim.Adam(parameter, lr=args.lr, weight_decay=args.weight_decay, eps=1e-8)
+
+
+# criterion = nn.CrossEntropyLoss()
+marginRankingLoss = nn.MarginRankingLoss(margin = 1, size_average = False)
+
+early_stop = False
+iterations = 0
+iters_not_improved = 0
+epoch = 0
+q2neg = {} # a dict from qid to a list of aid
+question2answer = {} # a dict from qid to the information of both pos and neg answers
+best_dev_map = 0
+best_dev_mrr = 0
+false_samples = {}
+
+start = time.time()
+header = '  Time Epoch Iteration Progress    (%Epoch)  Average_Loss Train_Accuracy Dev/MAP  Dev/MRR'
+dev_log_template = ' '.join(
+    '{:>6.0f},{:>5.0f},{:>9.0f},{:>5.0f}/{:<5.0f} {:>7.0f}%,{:>11.6f},{:>11.6f},{:12.6f},{:8.4f}'.split(','))
+log_template = ' '.join('{:>6.0f},{:>5.0f},{:>9.0f},{:>5.0f}/{:<5.0f} {:>7.0f}%,{:>11.6f},{:>11.6f},'.split(','))
+os.makedirs(args.save_path, exist_ok=True)
+os.makedirs(os.path.join(args.save_path, args.dataset), exist_ok=True)
+print(header)
+
+index2label = np.array(LABEL.vocab.itos)
+index2qid = np.array(QID.vocab.itos)
+index2aid = np.array(AID.vocab.itos)
+index2question = np.array(QUESTION.vocab.itos)
+index2answer = np.array(ANSWER.vocab.itos)
+
+
+# get the nearest negative samples to the positive sample by computing the feature difference
+def get_nearest_neg_id(pos_feature, neg_dict, distance="cosine", k=1):
+    dis_list = []
+    pos_feature = pos_feature.data.cpu().numpy()
+    pos_feature_norm = pos_feature / np.sqrt(sum(pos_feature ** 2))
+    neg_list = []
+    for key in neg_dict:
+        if distance == "l2":
+            dis = np.sqrt(np.sum((np.array(pos_feature) - neg_dict[key]["feature"]) ** 2))
+        elif distance == "cosine":
+            # feat_norm = normalize(np.array(neg_dict[key]["feature"].reshape(-1,1)), norm='l2')
+            neg_feature = np.array(neg_dict[key]["feature"])
+            feat_norm = neg_feature / np.sqrt(sum(neg_feature ** 2))
+            dis = 1 - feat_norm.dot(pos_feature_norm)
+        dis_list.append(dis)
+        neg_list.append(key)
+
+    k = min(k, len(neg_dict))
+    min_list = heapq.nsmallest(k, enumerate(dis_list), key=operator.itemgetter(1))
```

Comment:
You can merge `neg_list` and `dis_list` into `neg_dis_dict` and sort this dict by value to get top k id.

Reply:
Good suggestion. But the code will look almost the same and heap sort will be a little faster.